pb-rs/src/types.rs

```diff
@@ -1540,34 +1547,41 @@ impl FileDescriptor {
     ///
     /// Cycles means one Message calls itself at some point
     fn break_cycles(&mut self) {
-        let mut processing = Vec::with_capacity(self.messages.len());
+        let mut processing = VecDeque::with_capacity(self.messages.len());
         let mut messages = self
             .messages
             .iter()
             .map(|m| m.index.clone())
             .collect::<Vec<_>>();
         while let Some(m) = messages.pop() {
             messages.extend(m.get_message(self).messages.iter().map(|m| m.index.clone()));
-            processing.push((m.clone(), vec![m]));
+            processing.push_front((m.clone(), vec![m]));
         }
 
-        while let Some((i, ancestors)) = processing.pop() {
+        while let Some((i, ancestors)) = processing.pop_front() {
             let msg = i.get_message_mut(self);
+            let mut idx_first = 0;
+            debug!(
+                "break cycle:\nmessage: {:?}\nprocessing: {:?}",
+                i, processing
+            );
             for f in msg
-                .fields
-                .iter_mut()
-                .chain(msg.oneofs.iter_mut().flat_map(|o| o.fields.iter_mut()))
+                .all_fields_mut()
                 .filter(|f| !f.boxed && f.frequency != Frequency::Repeated)
```

Comment:
Does the test for `!f.boxed` do anything here? The only place we ever set `.boxed` is in this function, which visits each message only once, and visits each of its fields only once. So it seems to me `.boxed` should always be `false` at this point.

Reply:
So, I *think* this code is correct:

The `processing` queue contains exactly the set of not-yet-visited nodes. Visiting a node removes it from the queue forever, pulls its not-yet-visited direct successors to the front of the queue, and adds the path by which it was reached to their `ancestors` sets.

Because visiting a node pulls all its direct successors to the front of the queue, we visit nodes in a depth-first order: we always visit all not-yet-visited nodes reachable from that node before visiting any that are unreachable.

By the time the queue is empty, the entire graph is covered by one or more trees, in which each node's parent is the last of its predecessors to pull it to the front of the queue. Nodes never pulled to the front by any other are roots of their own trees.

Within one tree, each node's `ancestor` set contains all the nodes on the path to the root. Each node's parent is the node that pulled it to the front of the queue; and we always propagate the ancestors when we do so.

Every cycle in the graph has one node that was visited first, F. Since we visit every not-yet-visited node reachable from F before visiting any unreachable nodes, we will visit every node in the cycle before we visit any nodes unreachable from F. Hence, we will visit the node E that points to F. At this point, E's ancestor set will include F, and we will break the cycle.